### DISCUSSION TITLE
Experience with `next/root-params`

### DISCUSSION THREAD

**Discussion by amannn** [amannn/next-intl#1627] [General]
With [`next/root-params`](https://nextjs.org/docs/app/api-reference/functions/next-root-params) being available in Next.js 16.3, I'd be curious to hear how this feature is working for users of `next-intl`!

Here's a blog post I wrote on it:
→ [Using `next/root-params` in Next.js 16.3](https://next-intl.dev/blog/nextjs-root-params)

My impression so far has been very positive, these are my findings:

- Server Actions are not yet supported
- Route Handlers are not yet supported
- If you're using `createMiddleware` from `next-intl` in a proxy, be sure to split calls to `defineRouting` and `createNavigation` into separate files ([ref](https://github.com/amannn/next-intl/discussions/1627#discussioncomment-18066440))

Let me know if `next/root-params` works well for your app!

(this discussion initially tracked an earlier version of `next/root-params` but is now updated for Next.js 16.3)

**Comment by sannajammeh** [comment]
This API does not work for me currently in 15 latest canary. It simply returns {} for every page its called in

**Reply by amannn** [reply]
Do you by chance have a layout at `app/layout.tsx`? If so, that would cause the empty object. You need to make sure that you only have layouts in nested folders like `[locale]`.

**Comment by nevnein** [comment]
FYI this has been moved from `unstable_rootParams` to its own package(?), see https://github.com/vercel/next.js/releases/tag/v15.4.2-canary.42 and https://github.com/vercel/next.js/pull/82582

**Reply by amannn** [reply]
Yep, `unstable_rootParams` will be replaced by [`next/root-params`](https://github.com/vercel/next.js/pull/80255).

It's pretty much an equivalent for the previous API with one minor difference: Server Actions are (currently) not supported. AFAIK this is still under consideration and may change.

With Next.js 15.5, a deprecation notice for `unstable_rootParams` was added, but the API will continue to work until Next.js 16.

**Comment by GrahamQuan** [comment]
on next 15.5.0, current api can await a value, but `any` type

```ts
import { locale } from 'next/root-params';

const localeValue = await locale(); // `localeValue` type is any
```

i wonder how to add types in the future, `locale()` should be `Promise<'en' | 'es' | ...>`

**Reply by amannn** [reply]
I think it's coming: https://github.com/vercel/next.js/pull/80618

**Comment by jokull** [comment]
We just shipped triptojapan.com with root params. The code cleanup is great and we are finally using `generateStaticParams`. 

The only issue we ran into was in preview environments. In `generateStaticParams` we early returned an empty array, in combination with `export const dynamicParams = true;` to clamp down build times in preview environments. This resulted in errors for all routes with `generateStaticParams`, only preview envs but not in dev. Adding just a single prebuilt route with `slice(0, 1)` made it work.

**Reply by amannn** [reply]
Nice! 💯 



> This resulted in errors for all routes with `generateStaticParams`

That's odd! Thanks for sharing your workaround though!

**Comment by sannajammeh** [comment]
We also shipped https://nordicdownhillseries.com with root params. Works excellent!

**Comment by kachkaev** [comment]
What is the current state of this feature? I see a few related PRs from Summer 2025 (e.g. https://github.com/vercel/next.js/pull/82244) but it's been pretty quiet since then. Is Next.js still considering `next/root-params` the right abstraction or should we wait for something else instead? Mainly curious about root params in context of [`"use cache"` and i18n](https://aurorascharff.no/posts/implementing-nextjs-16-use-cache-with-next-intl-internationalization/).

**Reply by aidenlx** [reply]
At least in nextjs 16 release blog, they stated that an API is currently been worked on (probably mean `next/root-params`?) to replace `unstable_rootParams()` and will be shipped in an upcoming minor. 

https://nextjs.org/blog/next-16#breaking-changes-and-other-updates:~:text=unstable_rootParams()

**Comment by GrahamQuan** [comment]
it works great 🫶

git: https://github.com/GrahamQuan/nextjs-root-params-i18n

web: https://nextjs-root-params-i18n.vercel.app/

**Comment by Michaelkire** [comment]
I ran into an issue where I couldn't finish the implementation due to the following:

Your docs / blog tells to use the root param fetcher in the request config. However when I do, Nextjs yells at me, saying that it should be used in the app directory only. 

Not sure if it due to us using useExtracted. I will have to dig deeper.

**Reply by amannn** [reply]
Hmm, no—`useExtracted` shouldn’t make a difference here. Are you using the App Router or are you in some kind of migration state? Maybe your agent can isolate a minimal reproduction that demonstrates the issue so I can have a look!

**Reply by ClementDreptin** [reply]
I ran into a similar issue today. It took me a while but I think I found the cause.

In the basic example-app-router example, there is `src/i18n/routing.ts` that exports the routing configuration like so:
```ts
import {defineRouting} from 'next-intl/routing';

export const routing = defineRouting({
  locales: ['en', 'de'],
  defaultLocale: 'en',
  pathnames: {
    '/': '/',
    '/pathnames': {
      de: '/pfadnamen'
    }
  }
});
```
and `src/i18n/navigation.ts` that exports the wrappers around the `next/navigation` hooks and components like so:
```ts
import {createNavigation} from 'next-intl/navigation';
import {routing} from './routing';

export const {Link, getPathname, redirect, usePathname, useRouter} =
  createNavigation(routing);
```

In our app, `src/i18n/routing.ts` and `src/i18n/navigation.ts` are merged into a single file that looks like this:
```ts
import { createNavigation } from "next-intl/navigation";
import { defineRouting } from "next-intl/routing";

export const routing = defineRouting({
  locales: ["fr-FR", "en-GB", "es-ES"],
  defaultLocale: "fr-FR",
});

export const { Link, redirect, usePathname, useRouter } =
  createNavigation(routing);`
```

For whatever reason, the last 2 lines create the following crash:
```
Error: Invalid import
'next/root-params' can only be used inside the App Directory.
The error was caused by importing 'root-params.js'

Import trace:
  Middleware:
    [next]/root-params.js
    ...
```
Simply calling `createNavigation` without even exporting its result is enough to reproduce the crash.

To reproduce the crash on your end, simply run example-app-router with the following changes:
```diff
+import {createNavigation} from 'next-intl/navigation';
import {defineRouting} from 'next-intl/routing';

export const routing = defineRouting({
  locales: ['en', 'de'],
  defaultLocale: 'en',
  pathnames: {
    '/': '/',
    '/pathnames': {
      de: '/pfadnamen'
    }
  }
});
+
+createNavigation(routing);
```

I would be very curious to know why this happens, because the error message didn't help at all here.

**Reply by amannn** [reply]
What exactly is the error message you're seeing?

FWIW I switched the `next-intl` docs to use a separate file for `createNavigation` and `defineRouting` during the switch to Next.js 15, as that already showed issues: https://github.com/amannn/next-intl/issues/1669

Haven't tested having both calls in a single file in a longer time but would certainly recommend to separate them due to Next.js bundling behavior.

**Reply by ClementDreptin** [reply]
Here's the full error shown in the browser:
```
Error: Invalid import
'next/root-params' can only be used inside the App Directory.
The error was caused by importing 'root-params.js'

Import trace:
  Middleware:
    [next]/root-params.js
    ./src/i18n/request.ts
    ./node_modules/.pnpm/next-intl@4.13.6_@swc+helpers@0.5.23_next@16.3.1_@babel+core@7.29.0_supports-color@8.1._8b275c2cc7372963a1f783f3f9c70c09/node_modules/next-intl/dist/esm/development/server/react-server/getConfig.js
    ./node_modules/.pnpm/next-intl@4.13.6_@swc+helpers@0.5.23_next@16.3.1_@babel+core@7.29.0_supports-color@8.1._8b275c2cc7372963a1f783f3f9c70c09/node_modules/next-intl/dist/esm/development/navigation/react-server/getServerLocale.js
    ./node_modules/.pnpm/next-intl@4.13.6_@swc+helpers@0.5.23_next@16.3.1_@babel+core@7.29.0_supports-color@8.1._8b275c2cc7372963a1f783f3f9c70c09/node_modules/next-intl/dist/esm/development/navigation/react-server/createNavigation.js
    ./src/i18n/routing.ts
    ./src/proxy.ts
```

If having everything in the same file also causes bundle size issues, I have one more reason to keep things separate.

**Reply by amannn** [reply]
Ah yep, it seems like the bundling layer for the middleware is not considered "App Directory" for some reason.

But yeah, splitting the file is the way to go 👍

**Comment by om4rAb** [comment]
I am working with v 16.3.4 and following what nextjs Docs says about Internationalization but I always I get this error on next/root-params

**The export lang was not found in module [next]/root-params.js [app-rsc] (ecmascript).
The module has no exports at all.
All exports of the module are statically known (It doesn't have dynamic exports). So it's known statically that the requested export doesn't exist.** 

<img width="1222" height="727" alt="image" src="https://github.com/user-attachments/assets/52e1b305-85ab-43c6-990e-3e3cf3c000cd" />

**Reply by amannn** [reply]
Hmm, any chance you have a `app/layout.tsx`? Otherwise if you can extract a minimal reproduction, I'm happy to take a look.

**Reply by PacifiK2460** [reply]
Hi @amannn, I get the same error, currently my file structure is:

```
app:
- (app)
- (auth)
- ...
- layout.tsx
```

being `app/layout.tsx`

```tsx
import type { Metadata } from "next";
import "./globals.css";
import { Theme } from "@radix-ui/themes";
import { Toaster } from "sonner";


...
export default async function RootLayout({
    children,
}: Readonly<{
    children: React.ReactNode;
}>) {
    return (
        <html lang="es"
            className={`antialiased`}
        >
            <body
            >
                <Theme panelBackground="translucent" radius="full" accentColor="red">
                    <Toaster closeButton richColors />
                    {children}
                </Theme>
            </body>
        </html>
    );
}

```

and then `app/(app)/[lang]/layout.tsx`

```tsx

import { auth } from "@/lib/auth";
import { headers } from "next/headers";
import { redirect } from "next/navigation";

export default async function AppLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  const session = await auth.api.getSession({
    headers: await headers()
  });
  if (!session) {
    redirect("/login");
  }

  return (
    <>
      {children}
    </>
  );
}

```

I also have: `app/(app)/[lang]/page.tsx` for testing

```tsx
import { getDictionary } from "./dictionaries"

export default async function page() {
    const dict = await getDictionary()
    return (
        <>page; {dict.common.cancel}</>
    )
}
```

Where I have: `app/(app)/[lang]/dictionaries.tsx` 

```tsx
import { lang } from 'next/root-params'
import { notFound } from 'next/navigation'

const dictionaries = {
    en: () => import('./dictionaries/en.json').then((module) => module.default),
    es: () => import('./dictionaries/es.json').then((module) => module.default),
}

export type Locale = keyof typeof dictionaries

export const hasLocale = (locale: string): locale is Locale =>
    locale in dictionaries

export const getDictionary = async () => {
    const locale = await lang()
    if (!hasLocale(locale)) notFound()
    return dictionaries[locale]()
}

```
- I have 2 layouts so I can load all styles, currently my auth pages handle SSO, so literally no UI is rendered there
- The second layout does handles i8n, but the version I shared is the simple one

**Reply by PacifiK2460** [reply]
Hi @om4rAb and @amannn,

I managed to fix it by not using any Route Groups.

My currently working implementation follows the next structure:

```
app
    - [lang]
        - (app)
            - app
                - page.tsx
        - (auth)
            - ...
        - dictionaries.ts
        - globals.css
        - layout.tsx
    - page.tsx // Redirection page to default language (/es)
```

My guess is that root-params stops the param gathering once it encounters a Route Group?
If so, I think it may be worth that its mentioned in the [root-params page](https://nextjs.org/docs/app/api-reference/functions/next-root-params) or maybe in the [route groups page too](https://nextjs.org/docs/app/api-reference/file-conventions/route-groups).

---

Also @om4rAb, if you could share your minimal repo and we could help you :)

**Reply by amannn** [reply]
@PacifiK2460 The problem is your `app/layout.tsx`. I'd suggest to review https://next-intl.dev/blog/nextjs-root-params.

**Reply by om4rAb** [reply]
> Hmm, any chance you have a `app/layout.tsx`? Otherwise if you can extract a minimal reproduction, I'm happy to take a look.

Hi, I have been able to make it work, 
what I was doing wrong is, I did not have a root layout inside my group route, the moment I put the root layout inside my group route like this (see image) it worked very well 
<img width="233" height="254" alt="image" src="https://github.com/user-attachments/assets/6d2e7b0e-31cc-4d6f-b303-ecc89cbb3dc4" />
 
 just wanna mention that it raised an **generateStaticParam** error which can be solved by just addinng the following:
```
 export function generateStaticParams() {
	return [{ lang: "en" }, { lang: "fr" }]; 
}
```

One thing that I want you to help me with is now I should append the fr/en to every route and Im not sure how to do that?
giving a simple example just to clarify my issue, the moment I land on the home page **/fr/home** --> I click on products Link -->
it gets me to **/products** instead of **/fr/products** and same thing for every link


@PacifiK2460 this is what I did and it works i think the same like u did

**Reply by PacifiK2460** [reply]
@om4rAb, I would love to help you, but this may not be the best forum to do so.

Could you please send me an email? It's on my profile

**Reply by om4rAb** [reply]
ooh ok im sorry for that, will send u an email right away @PacifiK2460, thank you in advance

**Reply by om4rAb** [reply]
@amannn I would like to know if there's a way to apply i18n to just one side, see picture below to get my idea 

I wanna just apply the i18n to the **[lang]** and leave the **(admin)** out, doing so now gives me **404** **wen navigating to a route inside (admin)**

<img width="161" height="103" alt="image" src="https://github.com/user-attachments/assets/cefd5550-d0df-4e7e-9286-bf44e5da1e51" />

**Reply by amannn** [reply]
@om4rAb That's supported, yes. You'll want to backtrace where the 404 comes from. If you call APIs from `next-intl` in `(admin)`, you want to make sure that `getRequestConfig` can be called successfully if `rootParams.locale` returns `undefined`.

**Reply by om4rAb** [reply]
Hi @amannn sorry iv thoutgh that iv replied to u, I'm not using any third package for i18n i just went with what nextjs Docs says and it worked very well for so far, also for my last question I had to make two rootLayouts one for [lang]/(public) and another one for (admin) since I only want the public part to be translated 
so far everything is working well here's a record of the result, THANK YOU FOR YOUR HELP

record link:https://streamable.com/66sesv

**Reply by bashlk** [reply]
I wasn't using route groups and I ran into the same error. We are migrating an older project using pages router to app router and it wasn't immediately apparent from the docs that the root layout file needs to be within the root dynamic segment. Moving the root layout file within the root dynamic segment, in a similar fashion to what @om4rAb described above fixed the issue.

**Comment by daylilystudio** [comment]
Hi @amannn,
`next/root-params` has been working well for me overall,
but I ran into one issue I couldn't work around cleanly and figured it was worth reporting here.

**With `cacheComponents` enabled, a client-side navigation that changes the locale re-renders the
page but leaves the layout on the previous locale.**

Only happens in production — never in `next dev`, which makes sense since `<Link>` prefetching is
production-only, so every click in dev fetches fresh.

- Repo: https://github.com/daylilystudio/next-intl-cachecomponents-locale-repro
- Demo: https://next-intl-cachecomponents-locale-re.vercel.app

### Reproducing it on the demo

The layout prints its locale in a **red box**, each page prints its own in a **green box**. They
should always match.

1. Open https://next-intl-cachecomponents-locale-re.vercel.app — both boxes show `zh-TW` ✅
2. Click **`en`** in the header
3. URL becomes `/en`, green box switches to `en`, red box still says `zh-TW` ❌

Works the same going back (`en` → `zh-TW`) and on `/about`. A full page reload always renders
correctly, so it's only client-side navigation.

### What I checked

I assumed I'd misconfigured something, so I went through all the server output first — HTML, full
RSC response, and prefetch shell for `/`, `/en`, `/about`, `/en/about`. All correct. The switcher's
hrefs are right too, and the `/zh-TW/about` → `307 /about` + `set-cookie` chain works as designed.
So the server never sends the wrong locale.

The one asymmetry I found is that the segment-level prefetch tree leaves `[locale]` unresolved:

```json
"children": {
  "name": "locale",
  "param": { "type": "d", "key": null, "siblings": [] }
}
```

`"key": null`, which makes the tree for `/` and `/en` byte-identical — same md5, both 304 bytes on
the deployed demo. The full RSC response *does* carry `["locale","en",...]`, it's only this tree
that doesn't. Not confident enough about the internals to call that the cause, but it matches the
symptom: page segment updates, layout segment doesn't.

The repro follows your docs as closely as I could manage — official `<Link href={pathname}
locale={locale} />` switcher, locale resolved via root params in `i18n/request.ts`, no
`app/layout.tsx`, no route groups, and `defineRouting` / `createNavigation` in separate files
(thanks for that tip above).

### What I ended up doing

Plain `<a>` on the switcher so the browser does a full document load and skips the client router:

```tsx
<a href={getPathname({ href: pathname, locale, forcePrefix: true })} hrefLang={locale}>
```

`forcePrefix: true` turned out to matter — without it the locale cookie never gets written, since
you need to pass through the prefixed URL first. Took me a while to notice that's exactly what
`<Link locale>` was already doing for me.

Before landing on that I tried `instant = false` (no effect on the mismatch, though I was glad to
learn it doesn't make anything dynamic), `instantInsights.validationLevel` (no effect), and
`dynamicParams = false`, which fails the build as incompatible with `cacheComponents` — the exact
thing you opened [#84991](https://github.com/vercel/next.js/discussions/84991) about. Upvoted.

Also reassuring: `apostolos` in
[vercel/next.js#71927](https://github.com/vercel/next.js/discussions/71927) mentions they avoid the
related issues because they "manually do hard navigations on language change". Same conclusion,
arrived at independently.

**Reply by amannn** [reply]
Thank you for the detailed analysis! I just looked into this and it seems like this might be something to address on the Next.js side, I've opened an issue here: https://github.com/vercel/next.js/issues/98493.

As a workaround, it seems like you can `await params` in a segment—see the linked repro.

**Reply by daylilystudio** [reply]
Thanks for your help !!

**Reply by amannn** [reply]
https://github.com/vercel/next.js/pull/98674 seems to fix this
